### PR TITLE
Improve performance on event filtering

### LINKED ISSUE #19142: Improve performance of splitters in vector with FilterEvents
It is found that  for algorithm FilterEvents, the computation speeds to filter events are different between using SplittersWorkspace and MatrixWorkspace.  

### Expected behavior

The speed to use MatrixWorkspace should be faster or at least same than using SplittersWorkspace.

### Actual behavior

For large EventWorkspace, such as NOMAD, the speed to filter by MatirxWorkspace splitters is twice as long as that by SplittersWorkspace.

### Platforms affected

All.

### PR DESCRIPTION
**Pull request by wdzhou**
Improve performance on event filtering

**To test:**

1. Check the unit tests and doc tests.
2. Create 3 identical splitters workspace in SplittersWorkspace, MatrixWorkspace and TableWorkspace.
3. Use them to filter a same unsorted EventWorkspace an check whether the results are identical.
4. Use them with SNSPowderReduction and check whether the outputs are identical. 
5. Check the documents of FilterEvents.

Fixes #19142.

*New features are reported in docs/source/release/v3.10.0/framework.rst*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by wdzhou** [comment]
Next to implement:
1. Add sortEvent in FilterEvents()
2. Clean the part of code for splitting log

**Comment by wdzhou** [comment]
Next to do: 
1. Clean the part of code for splitting log in FilterEvents
2. In SplitMatrixHelper, stop the filtering if there is no splitter left
3. Release note

File: docs/source/algorithms/FilterEvents-v1.rst [hunk 1]
```diff
@@ -20,19 +20,48 @@ Inputs
...
+Event splitter
+++++++++++++++
+
+An event splitter contains three items, start time, stop time and splitting target (index).
+All the events belonged to the same splitting target will be saved to a same output EventWorkspace.
+
+Unit of input splitters
++++++++++++++++++++++++
+
+ - MatrixWorkspace:  the unit must be second.
```
Comment (peterfpeterson) [inline docs/source/algorithms/FilterEvents-v1.rst]: Get rid of the indentation of the bullet points and the empty line between them

**Review by peterfpeterson — DISMISSED** [review]
The unordered list at L131 should move up one level (delete a space on the left).

All references to properties should be switched to escaped code (with double back-ticks) rather than bold (double asterisks).

Misspelling of `filter evnets` on line 197.

**Comment by peterfpeterson** [comment]
Overall it works as advertised. There are a few changes to fix up the sphinx pages which are all fairly trivial.

**Comment by peterfpeterson** [comment]
:shipit:

**Pull request merged**